### DISCUSSION TITLE
Modify input parameter 'NAME' during runtime [Control]

### DISCUSSION THREAD

**Discussion by Traiwit** [idaholab/moose#19167] [Q&A General] [answered]
Hi guys,

I'm wondering if it's possible to modify input parameter 'NAME' during runtime?

Since my simulation has quite a number of steps (100-300 mining frames), at each step we need to upload material property (damage) and it becomes very messy and might affect the memory usage of the simulation

for example at each time step we need to have 4 objects

```
userobject block
  [./damage_read_1]
    type = ElementPropertyReadFile
    prop_file_name = 'damage_1.csv'     ******** control name here
    nprop = 1
    read_type = element
  [../]

material block

  [damage_1]
    type = GenericConstantArrayTC
    prop_name = damage_1
    read_prop_user_object = damage_read_1   ******** or here
    []

  [./permeability_all_1]
    type = PorousFlowPermeabilityConstTC
    read_prop_user_object = perm_read
    damage = damage_1
    aw = 20
    kmax = 1e-7
  [../]
  
Control block
  [right_control]
    type = ConditionalFunctionEnableControl
    enable_objects = 'Materials::permeability_all_1'
    conditional_function = ''if(t=2,1,0)''
    execute_on = 'initial timestep_begin'
  []
```

instead, is it possible to just change the highlighted parameter name, base on the list from .csv file using [Control]?

in this case, I can just change the `material/damage/read_prop_user_object`  then everything can stay the same.

Or even better at `userobject/ElementPropertyReadFile/prop_file_name = 'damage_1.csv'`

for example 
```
t1 userobject/ElementPropertyReadFile/prop_file_name = 'damage_1.csv'
t2 userobject/ElementPropertyReadFile/prop_file_name = 'damage_2.csv'
t3 userobject/ElementPropertyReadFile/prop_file_name = 'damage_3.csv'
```

in this case, i'd only need 100-300 control objects rather than 4x(100-300) objects in the input file.

I checked https://mooseframework.inl.gov/syntax/Controls/index.html
but not sure which one is the right one for me

![image](https://user-images.githubusercontent.com/79562439/138196707-1471de26-0fe7-44a0-889a-1c0cb9bfd1cd.png)

Kind regards,
Traiwit

**Comment by GiudGiud** [comment, marked as answer]
Not really. You can control objects, Real and bool. For some objects it would be totally valid to change a string, but for many others the string parameter is used to fetch a reference at initialization and changing the string during the simulation would just not do anything. 

I think your solution here is to code the reading of a CSV file with a column for the time coordinate.

**Reply by GiudGiud** [reply]
another solution would be to actually control the string (so add a new Control)
and to add a timestepSetup routine to the UserObject (or whichever system you want the Control to intervene on) and in that routine, query the right object after the control has run

The timestepSetup routine comes from the SetupInterface, make sure the object you want to use inherits that interface

**Reply by Traiwit** [reply]
Hi @GiudGiud, thank you! i'm still not sure if I understand all that yet, I need some time to digest and I will get back to you soon.

**Reply by Traiwit** [reply]
@GiudGiud, also a quick question, apparently it seems like my own material isn't controllable

The following error occurred in the object "damage_control_1", of type "ConditionalFunctionEnableControl".

this is the error message:

`The desired parameter 'Materials::damage1/enable' was not located for the 'damage_control_1' object, it either does not exist or has not been declared as controllable.`

I thought it'd work out of the box, `BCs` and `UserObjects` are controllable though, not sure why materials isn't, where should I look into this?

Kind regards,
Traiwit

**Reply by GiudGiud** [reply]
How old is your MOOSE?
we have fixed that (Materials being controllable) a few weeks ago

**Reply by Traiwit** [reply]
> How old is your MOOSE? we have fixed that (Materials being controllable) a few weeks ago

Hi @GiudGiud I updated it, seems like that problem is gone, but doesn't seem like it works properly

so I tested it with MOOSE default material

```
Control block
  [perm_1]
      type = ConditionalFunctionEnableControl
      enable_objects = 'Materials::permeability_1'
      conditional_function = 'if(t>=0 & t<1,1,0)'
      execute_on =  'initial timestep_begin'
  []
  [perm_2]
    type = ConditionalFunctionEnableControl
    enable_objects = 'Materials::permeability_2'
    conditional_function = 'if(t>=1 & t<2,1,0)'
    execute_on =  'initial timestep_begin''
  []

Materials block
    [permeability_1]
    type = PorousFlowPermeabilityConst
    permeability = '1e-9 0 0 0 1e-9 0 0 0 1e-9'
   []
  [permeability_2]
  type = PorousFlowPermeabilityConst
  permeability = '5e-9 0 0 0 5e-9 0 0 0 5e-9'
 []
```

So after T>=1, permeability_1 should be deactivate, but seems like it's still active

So i got this error right at the start of the simulation

```
*** ERROR ***
The following error occurred in the object "MOOSE Problem", of type "FEProblem".

The following material properties are declared on block 2 by multiple materials:
  Material Property             Material Objects
  PorousFlow_permeability_qp    PorousFlow_density_nodal_all PorousFlow_density_qp_all PorousFlow_relative_permeability_nodal_all PorousFlow_relative_permeability_qp_all PorousFlow_viscosity_nodal_all PorousFlow_viscosity_qp_all permeability_1 permeability_2 porosity_nodal 
  dPorousFlow_permeability_qp_dgradvarPorousFlow_density_nodal_all PorousFlow_density_qp_all PorousFlow_relative_permeability_nodal_all PorousFlow_relative_permeability_qp_all PorousFlow_viscosity_nodal_all PorousFlow_viscosity_qp_all permeability_1 permeability_2 porosity_nodal 
  dPorousFlow_permeability_qp_dvarPorousFlow_density_nodal_all PorousFlow_density_qp_all PorousFlow_relative_permeability_nodal_all PorousFlow_relative_permeability_qp_all PorousFlow_viscosity_nodal_all PorousFlow_viscosity_qp_all permeability_1 permeability_2 porosity_nodal 
```

Moreover, when I removed `Control/perm2` and `Materials/permeability_2` the simulation still works fine even when T>1, but perm1 should be gone when T>1. This makes me think the permeability_1 is still there even I deactivate it.

Could you please help me look into this?

Kind regards,
Traiwit

**Reply by Traiwit** [reply]
Hi @GiudGiud all good now, I just need

```
[Problem]
  material_dependency_check = false
[]
```

in my input file, not sure why, but it works, I will test with an actual case and get back to you

**Reply by GiudGiud** [reply]
merge the two control blocks and have a disable_objects parameter, i think that will help

**Reply by Traiwit** [reply]
> merge the two control blocks and have a disable_objects parameter, i think that will help

I was thinking of doing that using `TimePeriod`, but when I have 120+ objects, I'm not sure how to chain them

rn I'm doing the simplest case where I have 120 control objects and 120 material objects

**Reply by GiudGiud** [reply]
Ah yes.
Well if it works with 120 controls I m glad but you could just write a single control that handles a vector of inputs and only activates one at a time at this point.

**Reply by Traiwit** [reply]
> Ah yes. Well if it works with 120 controls I m glad but you could just write a single control that handles a vector of inputs and only activates one at a time at this point.

yeh that's a good idea (similar to my constraints, I reckon), at this point I'm still not sure how to implement it with time, I know you posted about it above, but still not sure where to look for

**Reply by GiudGiud** [reply]
Well if you're controlling the activation of materials (rather than controlling the string parameter inside a single one) then you can take a look at https://github.com/idaholab/moose/blob/next/framework/src/controls/ConditionalEnableControl.C

This part of the control system is pretty intuitive imo with regards to APIs, just an enable parameter to switch on and off
